File: sort-mapper-example-java/src/main/java/kh/farrukh/sortmapperexamplejava/course/CourseController.java
package kh.farrukh.sortmapperexamplejava.course;

import kh.farrukh.sortmapperexamplejava.course.model.CourseCreateRequestDTO;
import kh.farrukh.sortmapperexamplejava.course.model.CourseDetailsResponseDTO;
import kh.farrukh.sortmapperexamplejava.course.model.CourseResponseDTO;
import kh.farrukh.sortmapperexamplejava.course.model.CourseUpdateRequestDTO;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequiredArgsConstructor
public class CourseController implements CourseApi {

  private final CourseService courseService;

  @Override
  public ResponseEntity<Page<CourseResponseDTO>> getCourses(Pageable pageable) {
    return ResponseEntity.ok(courseService.getCourses(pageable));
  }

  @Override
  public ResponseEntity<CourseDetailsResponseDTO> getCourse(Long id) {
    return ResponseEntity.ok(courseService.getCourse(id));
  }

  @Override
  public ResponseEntity<Void> createCourse(CourseCreateRequestDTO requestDTO) {
    courseService.createCourse(requestDTO);
    return ResponseEntity.status(HttpStatus.CREATED).build();
  }

  @Override
  public ResponseEntity<Void> updateCourse(Long id, CourseUpdateRequestDTO requestDTO) {
    courseService.updateCourse(id, requestDTO);
    return ResponseEntity.ok().build();
  }

  @Override
  public ResponseEntity<Void> deleteCourse(Long id) {
    courseService.deleteCourse(id);
    return ResponseEntity.noContent().build();
  }

}

File: sort-mapper-example/src/main/kotlin/kh/farrukh/sortmapperexample/course/CourseController.kt
package kh.farrukh.sortmapperexample.course

import kh.farrukh.sortmapperexample.course.model.CourseCreateRequestDTO
import kh.farrukh.sortmapperexample.course.model.CourseShortInfoResponseDTO
import kh.farrukh.sortmapperexample.course.model.CourseUpdateRequestDTO
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.data.web.PagedModel
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.RestController

@RestController
class CourseController(
    private val courseService: CourseService
) : CourseApi {

    override fun getCourses(pageable: Pageable) =
        ResponseEntity.ok(PagedModel(courseService.getCourses(pageable)))

    override fun getCoursesShortInfo(sort: Sort): ResponseEntity<List<CourseShortInfoResponseDTO>> {
        return ResponseEntity.ok(courseService.getCoursesShortInfo(sort))
    }

    override fun getCourseById(id: Long) = ResponseEntity.ok(courseService.getCourseById(id))

    override fun createCourse(requestDTO: CourseCreateRequestDTO): ResponseEntity<Unit> {
        courseService.createCourse(requestDTO)
        return ResponseEntity.status(HttpStatus.CREATED).build()
    }

    override fun updateCourse(id: Long, requestDTO: CourseUpdateRequestDTO): ResponseEntity<Unit> {
        courseService.updateCourse(id, requestDTO)
        return ResponseEntity.ok().build()
    }

    override fun deleteCourse(id: Long): ResponseEntity<Unit> {
        courseService.deleteCourse(id)
        return ResponseEntity.noContent().build()
    }

}